From cd3a75cd1efea5defcfb9c349b5328182b34dc4d Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Sun, 12 Dec 2021 08:45:36 -0500
Subject: [PATCH 05/10] Mark accelerometer input as sensor exclusively if there
 are ABS axis

The reason this is needed is that on -fucked up- Xiami Mi A2 Lite and
Galaxy J8, the input device reporting power button/vol- reports ACCELEROMETER
input prop.
This leads to wrong timestamping ioctl, leading to crashes and broken UI

Change-Id: I491ab89724c908ef09a23a6427d24a8889bf806a
---
 services/inputflinger/reader/EventHub.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/services/inputflinger/reader/EventHub.cpp b/services/inputflinger/reader/EventHub.cpp
index d6a6bd214e..febce0ae6b 100644
--- a/services/inputflinger/reader/EventHub.cpp
+++ b/services/inputflinger/reader/EventHub.cpp
@@ -2124,7 +2124,13 @@ void EventHub::openDeviceLocked(const std::string& devicePath) {
 
     // Check whether this device is an accelerometer.
     if (device->propBitmask.test(INPUT_PROP_ACCELEROMETER)) {
-        device->classes |= InputDeviceClass::SENSOR;
+        bool hasAxis = false;
+        for (int i = 0; i <= ABS_MAX; i++) {
+            if (device->absBitmask.test(i)) hasAxis = true;
+        }
+        if(hasAxis) {
+            device->classes |= InputDeviceClass::SENSOR;
+        }
     }
 
     // Check whether this device has switches.
-- 
2.25.1

